### ISSUE TITLE
Screen flickering on iOS with & without Impeller enabled 

### ISSUE THREAD

**Issue by delfme** [flutter/flutter#128221]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

@zanderso We woke up with a bad surprise.
After upgrading to latest master 3.11.0-18.0.pre.76 this flickering issue is occurring consistently in our app.
No errors from console. 

Tested on ios 16.1, iphone 13 pro.

### Expected results

Screen not flickering.

### Actual results

Please see video attached in below section.

[UPDATED]
A Widget, which is displaying an instagram-like story, gets closed by a swipe-down gesture.
It seems that this is due to a conflict between latest master and `ExtendedNetworkImageProvider` provider.

Notes:
- We've never experienced this issue before, so this must be a regression occurred with a change on master during 1-2 months
- Issue occurs even if no Platform view is on the scene.
- Issue occurs consistently, but no error is thrown by console. Here is timeline: https://drive.google.com/file/d/1WJKpqR0RABBM64xHRMfOp7FuX10q07v5/view?usp=sharing

### Code sample
This causes flickering if it is called when we are "close" to a transition.

```
precacheImage(
      ExtendedNetworkImageProvider(
           image_url,
           cache: true,
           cacheMaxAge: const Duration(days: 2),
      ),
      context,
 );
```

### Screenshots or Video

Here is a video reproducing issue:

https://github.com/flutter/flutter/assets/53510751/fb4b297c-e722-4706-9697-dd50516652dd

### Logs

Console throws no errors when issue occurs.

### Flutter Doctor output

[✓] Flutter (Channel master, 3.11.0-18.0.pre.76, on macOS 12.6.5 21G531 darwin-arm64, locale en-IT)
[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.2)
[✓] Connected device (3 available)
[✓] Network resources

**Comment by luckysmg** [comment]
Not very sure. But from the behavior I think maybe it is related with partial repaint.
You can set `FLTDisablePartialRepaint` to `YES` in your `Info.plist` to have a try.

**Comment by delfme** [comment]
Found out that the culprit is a conflict between latest master and ExtendedImage.
Basically flickering is triggered by precaching image with `ExtendedNetworkImageProvider`
Point here the snippet, since this is a much used plugin

```
precacheImage(
                  ExtendedNetworkImageProvider(
                    image_url,
                    cache: true,
                    cacheMaxAge: const Duration(days: 2),
                  ),
                  context,
                );
```

**Comment by delfme** [comment]
> Not very sure. But from the behavior I think maybe it is related with partial repaint.
> You can set `FLTDisablePartialRepaint` to `YES` in your `Info.plist` to have a try.

I tried it and it doesn't fix issue in that case.
We temporarily disabled precaching and opened an issue here https://github.com/fluttercandies/extended_image/issues/585

**Comment by delfme** [comment]
Also writing a quick workaround.

```
 WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      Future.delayed(const Duration(milliseconds: 2000), () async {
          // start precaching media far away from transition animation
     });
 }); 
```

**Comment by delfme** [comment]
Reopening this coz it seems issue might not be due to plugin 
https://github.com/fluttercandies/extended_image/issues/585

I have no knowledge to judge this, it might be worth for flutter team having a look at it.

**Comment by dam-ease** [comment]
Hi @delfme. Thanks for filing this issue. Kindly provide a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) to help investigate this.

**Comment by delfme** [comment]
It is hard for me to work on that at the moment. I guess a demo should just run that precache code while a transition is occurring or soon after (maybe during a screen animation?). There is something that happens which Impeller engine cannot deal with and the flickering kicks off.

**Comment by delfme** [comment]
But it is happening everywhere here. We need to find a workaround. ExtendedImage is much used, please play a bit around with it to see issue. I recall it was one of the culprits mentioned inside the infamous https://github.com/flutter/flutter/issues/100522
Maybe there is still something conflicting with it. Plugin authors believe it is just Impeller and they can do nothing about it.
Also, regression has occurred recently (<2months), so it should be easier to find a cause investigating it now.

**Comment by delfme** [comment]
@zanderso will try to do that 👍

Got an idea. Not orthodox but should work.
If images from Wonderous app were moved to aws s3 or somewhere on the cloud and precached with ExtendedImageProvider, I think we are very likely to see flickering issues
https://github.com/gskinnerTeam/flutter-wonderous-app

I never compiled that app but saw it has images + animations. It would be a valid sample app to test these kind of issues against to.

**Comment by delfme** [comment]
But regression seems to involve images at large.

For example, just noticed we get this error inside our search screen where we return some avatars.

```
======== Exception caught by image resource service ================================================
The following ArgumentError was thrown resolving an image codec:
Invalid argument(s): No host specified in URI file:///

When the exception was thrown, this was the stack: 
#0      _HttpClient._openUrl (dart:_http/http_impl.dart:2751:9)
#1      _HttpClient.getUrl (dart:_http/http_impl.dart:2627:48)
#2      NetworkImage._loadAsync (package:flutter/src/painting/_network_image_io.dart:124:59)
#3      NetworkImage.loadImage (package:flutter/src/painting/_network_image_io.dart:84:14)
#4      ImageProvider.resolveStreamForKey.<anonymous closure> (package:flutter/src/painting/image_provider.dart:529:39)
#5      ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:381:22)
#6      ImageProvider.resolveStreamForKey (package:flutter/src/painting/image_provider.dart:526:81)
#7      ImageProvider.resolve.<anonymous closure> (package:flutter/src/painting/image_provider.dart:374:9)
#8      ImageProvider._createErrorHandlerAndKey.<anonymous closure> (package:flutter/src/painting/image_provider.dart:488:24)
#9      SynchronousFuture.then (package:flutter/src/foundation/synchronous_future.dart:41:39)
#10     ImageProvider._createErrorHandlerAndKey (package:flutter/src/painting/image_provider.dart:485:9)
#11     ImageProvider.resolve (package:flutter/src/painting/image_provider.dart:371:5)
#12     DecorationImagePainter.paint (package:flutter/src/painting/decoration_image.dart:310:55)
#13     _BoxDecorationPainter._paintBackgroundImage (package:flutter/src/painting/box_decoration.dart:471:20)
#14     _BoxDecorationPainter.paint (package:flutter/src/painting/box_decoration.dart:488:5)
#15     RenderDecoratedBox.paint (package:flutter/src/rendering/proxy_box.dart:2291:17)
#16     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#17     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#18     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:129:13)
#19     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#20     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#21     RenderBoxContainerDefaultsMixin.defaultPaint (package:flutter/src/rendering/box.dart:2878:15)
#22     RenderFlex.paint (package:flutter/src/rendering/flex.dart:1040:7)
#23     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#24     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#25     RenderBoxContainerDefaultsMixin.defaultPaint (package:flutter/src/rendering/box.dart:2878:15)
#26     RenderFlex.paint (package:flutter/src/rendering/flex.dart:1040:7)
#27     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#28     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#29     RenderShiftedBox.paint (package:flutter/src/rendering/shifted_box.dart:74:15)
#30     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#31     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#32     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:129:13)
#33     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#34     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#35     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:129:13)
#36     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#37     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:253:13)
#38     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:129:13)
#39     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:3059:7)
#40     PaintingContext._repaintCompositedChild (package:flutter/src/rendering/object.dart:169:11)
#41     PaintingContext.repaintCompositedChild (package:flutter/src/rendering/object.dart:112:5)
#42     PipelineOwner.flushPaint (package:flutter/src/rendering/object.dart:1147:31)
#43     RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:494:19)
#44     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:918:13)
#45     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:358:5)
#46     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1297:15)
#47     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1227:9)
#48     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1085:5)
#49     _invoke (dart:ui/hooks.dart:170:13)
#50     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:400:5)
#51     _drawFrame (dart:ui/hooks.dart:140:31)
Image provider: NetworkImage("", scale: 1.0)
Image key: NetworkImage("", scale: 1.0)
```

Ofc, URI parsing is not the issue, which is instead related to ImageProvider or Image.

Infact, if we just disable the image widget from the screen, the error goes away.

Here is screen with and without Images, images render but console throws above error.

https://github.com/flutter/flutter/assets/53510751/9772187b-825a-4981-82e1-fd8a2d72d5da


https://github.com/flutter/flutter/assets/53510751/81ca6d1b-cc1a-4949-a014-3883a90b630d

The culprit is this simple code:
```
return ExtendedImage.network(
  imageUrl,
  cache: true,
  cacheMaxAge: const Duration(days: 2),
  height: widget.size,
  width: widget.size,
  fit: BoxFit.cover,
…
```

Issue resolves if we substitute above code with
`return Image.network(imageUrl);`

Whatever the issue is, I believe it is also causing the flickering, coz flickering is triggered by this:

```
precacheImage(
      ExtendedNetworkImageProvider(
         …
)
```

CC: @zmtzawqlp are you sure it is all on Impeller side? We indeed got these issues only after upgrading to latest master but maybe plugin is also involved somehow?

**Comment by zmtzawqlp** [comment]
@delfme need a simple runnable demo to reproduce it

**Comment by dam-ease** [comment]
Hi @delfme. Thanks for the detailed explanation.
Did you try without Imepeller to see if the issue still persists? 
Also, it would be helpful if you can narrow down per this [comment](https://github.com/flutter/flutter/issues/128221#issuecomment-1579001087)

For issues to be workable, it needs to be reproducible with a minimal complete code sample that doesn’t include 3rd party plugins or complex production code.

**Comment by delfme** [comment]
@zmtzawqlp I will try. 
For now I quick got to see that issue occurs also when using `CachedNetworkImage`. 
```
 return CachedNetworkImage(
            imageUrl: imageUrl,
            progressIndicatorBuilder: (context, url, downloadProgress) =>
                CircularProgressIndicator(value: downloadProgress.progress),
            errorWidget: (context, url, error) => Icon(Icons.error),
          );
```

Unless both plugins have same bug, that makes us think it might truly be something at flutter engine level.

@dam-ease @zanderso Unfortunately that bug happened in the wrong moment. Will try to make time for this. Meantime my idea here is worth a trying if you already use this app for internal testing. It makes sense to add remote images inside a fork of the app so that the more complex use case of dealing with remote images can be checked too. 
https://github.com/flutter/flutter/issues/128221#issuecomment-1579205182

I tested without Impeller and issue is still reproducible.
As for the plugin, I guess `CachedNetworkImage` is flutter default solution.

**Comment by delfme** [comment]
@dam-ease thx! Will do it. Want also to better check if it is something in our code. I just noticed there was a little change since issue occurred. Console not throwing errors doesnt help.

**Comment by dam-ease** [comment]
Ok, looking forward to your response.

**Comment by delfme** [comment]
Im debugging, it is hard to nail down. At glance seems like getting `context` inside build() + some transition + `precacheImage` together trigger flickering

**Comment by dam-ease** [comment]
Hi @delfme. Kindly share your debugging results once you are done debugging and you have a reproducible code sample to share.

**Comment by darshankawar** [comment]
@delfme You can share your latest findings in this issue.

**Comment by delfme** [comment]
Thx @darshankawar.

Here is latest findings, hope this helps:

I added a log to didHaveMemoryPressure to see if images flickering was somehow linked to underneath cache activity and I think I nailed down the issue.
It seems that flickering occurs rigth when didHaveMemoryPressure is being invoked and log
"Warning!! Low Memory, didHaveMemoryPressure" is printed on console. When log is not printed, I see no flickering and viceversa when flickering occurs, the log is also printed.

My best guess with this is that flutter engine frees up image cache while at same time it is loading images (see code paragraph). Issue occurs with both ExtendedNetworkImage and CachedNetworkImage. It is like by cleaning data while images are getting rendered, engines finds itself rendering the very data it just cleaned and flickering occurs.

**Comment by zanderso** [comment]
cc @dnfield

**Comment by dnfield** [comment]
When the engine gets a memory pressure signal it also evicts Skia's caches aggressively.

The image cache should not really cause the whole screen to flicker, but maybe something with Skia's resource cache could. 

@delfme can you clarify a bit whether this memory pressure issue also causes flickering with Impeller? With Impeller we don't do anything special with low memory notifications. However, when we get them we _do_ ask Dart to do a pretty hefty blocking GC ASAP. Can you share a trace that includes the flickering behavior? We should be able to see the GC activity in the trace (the one linked at the very top of this bug doesn't seem to have too much in it related to that).

**Comment by dnfield** [comment]
@delfme you seem to have a different bug than what's being reported in this issue - the bug in this issue was the _entire_ screen flickering at once, not images in the scene IIRC.

**Comment by delfme** [comment]
Prolly the first video of entire screen flickering occurred when we tested issue with skia.

We can change the issue title since we use Impeller, although things could be intertwined somehow.

**Comment by delfme** [comment]
Screen flickering just occurred with Impeller as well.

So, to recap, issue is both with images flickering or screen flickering.
Screen flickering occurs more randomdly then images flickering but still occurs.

**Comment by delfme** [comment]
@darshankawar I asked guys here to check if iusse is related to our discovery. If flickering is due to cache cleaning during images rendering, IG also other pending flickering issues can resolve after fix.

https://github.com/flutter/flutter/issues/105093

**Comment by dnfield** [comment]
I confused this with another issue.

This looks like some kind of bug with the Extended image package. 

FWIW, implementing an image widget/render object is particularly challening. You have to get a bunch of things right around interactions between the image cache and how it tracks various types of images ("live" vs "pending" vs "cached").

If your app is getting low memory signals, it seems quite likely that you are simply caching too many images - you need to cache fewer  if you want to target the device you're hitting. A flickering animation when the image reloads is expected _especially if your image widget implementation is not correctly interacting with the live images part of the cache_.

I'm going to close this on the assumption that it is a bug in the package you linked to. If you can provide a small example of those using framework only API then I'd be happy to reopen it. Alternatively, if you can bisect to a commit where framework/engine regressed behavior in the package we could try to look at that. As it is, the Flutter team is not going to invest time into reading the package code to see where it might be doing something unexpected.

**Comment by delfme** [comment]
Hi Dan, I also was wondering whether this was a plugin bug and thought it was engine bug only after seeing that bug occurs on both `ExtendedImage` and `CachedNetorkImage`. But ofc that was not enough to rule out the culprit.

Since this is challenging and also much needed feature, could flutter team consider to develop an image cache solution or have a second thought on the possibility of you guys collab with available plugins? Thx

**Comment by delfme** [comment]
Just informing @dnfield guess was right. Hiding above comment as outdated.
